ospf6d/ospf6_asbr.c

```diff
@@ -2613,3 +2693,990 @@ void install_element_ospf6_debug_asbr(void)
+	route_tag_t tag = 0;
+	int mtype;
+
+	lsa = ospf6_lsdb_lookup(
+		htons(OSPF6_LSTYPE_AS_EXTERNAL),
+		htonl(aggr->id), ospf6->router_id,
+		ospf6->lsdb);
+	if (!lsa) {
+		zlog_warn(
+			"%s: Could not refresh/originate %pFX",
+			__func__,
+			&aggr->p);
+
+		return OSPF6_FAILURE;
+	}
+
+	asel = (struct ospf6_as_external_lsa *)
+		OSPF6_LSA_HEADER_END(lsa->header);
+	metric = (unsigned long)OSPF6_ASBR_METRIC(asel);
+	tag = ospf6_as_external_lsa_get_tag(lsa);
+	mtype = CHECK_FLAG(asel->bits_metric,
+			   OSPF6_ASBR_BIT_E) ? 2 : 1;
+
+	/* Fill all the details for advertisement */
+	ospf6_fill_aggr_route_details(ospf6, aggr);
+	rt_aggr = aggr->route;
+	/* If tag/metric/metric-type modified , then
+	 * re-originate the route with modified
+	 * tag/metric/metric-type details.
+	 */
+	if ((tag != aggr->tag)
+	    || (metric
+		!= (unsigned int)rt_aggr->path.cost)
+	    || (mtype
+		!= aggr->mtype)) {
+		if (IS_OSPF6_DEBUG_AGGR)
+			zlog_debug(
+			"%s: Changed tag(old:%d new:%d)/metric(o:%u n:%d)/mtype(o:%d n:%d),So refresh the summary route.(%pFX)",
+			__func__, tag,
+			aggr->tag,
+			metric,
+			(unsigned int)rt_aggr->path.cost,
+			mtype, aggr->mtype,
+			&aggr->p);
+
+		(void)ospf6_originate_type5_type7_lsas(
+					aggr->route,
+					ospf6);
+	}
+
+	return OSPF6_SUCCESS;
+}
+
+static void ospf6_handle_external_aggr_update(struct ospf6 *ospf6)
+{
+	struct route_node *rn = NULL;
+	int ret;
+
+	if (IS_OSPF6_DEBUG_AGGR)
+		zlog_debug("%s: Process modified aggregators.", __func__);
+
+	for (rn = route_top(ospf6->rt_aggr_tbl); rn; rn = route_next(rn)) {
+		struct ospf6_external_aggr_rt *aggr;
+
+		if (!rn->info)
+			continue;
+
+		aggr = rn->info;
+
+		if (aggr->action == OSPF6_ROUTE_AGGR_DEL) {
+			aggr->action = OSPF6_ROUTE_AGGR_NONE;
+			ospf6_asbr_summary_config_delete(ospf6, rn);
+
+			if (OSPF6_EXTERNAL_RT_COUNT(aggr))
+				hash_clean(aggr->match_extnl_hash,
+				(void *)ospf6_aggr_handle_external_info);
+
+			hash_free(aggr->match_extnl_hash);
+			XFREE(MTYPE_OSPF6_EXTERNAL_RT_AGGR, aggr);
+
+		} else if (aggr->action == OSPF6_ROUTE_AGGR_MODIFY) {
+
+			aggr->action = OSPF6_ROUTE_AGGR_NONE;
+
+			/* Check if tag/metric/metric-type modified */
+			if (CHECK_FLAG(aggr->aggrflags,
+				OSPF6_EXTERNAL_AGGRT_ORIGINATED)
+			    && !CHECK_FLAG(aggr->aggrflags,
+				OSPF6_EXTERNAL_AGGRT_NO_ADVERTISE)) {
+
+				ret = ospf6_handle_external_aggr_modify(ospf6,
+									aggr);
+				if (ret == OSPF6_FAILURE)
+					continue;
+			}
+
+			/* Advertise option modified ?
+			 * If so, handled it here.
+			 */
+			ospf6_aggr_handle_advertise_change(ospf6, aggr);
+		}
+	}
+}
+
+static void ospf6_aggr_unlink_external_info(void *data)
+{
+	struct ospf6_route *rt = (struct ospf6_route *)data;
+
+	rt->aggr_route = NULL;
+
+	rt->to_be_processed = true;
+}
+
+void ospf6_external_aggregator_free(struct ospf6_external_aggr_rt *aggr)
+{
+	if (OSPF6_EXTERNAL_RT_COUNT(aggr))
+		hash_clean(aggr->match_extnl_hash,
+			(void *)ospf6_aggr_unlink_external_info);
```

Comment:
This cast is not necessary. The function prototype matches the function call signature.